### PR TITLE
Update Exterminate

### PR DESCRIPTION
**Pull request by Crystalwarrior**
* Fix it crashing adventure mode due to outdated item deletion method
* Add --x (--max) argument, letting you set an extermination cap
* Teleport disintegrated units to the top of the world so they are not seen dying

### CODE REVIEW

File: exterminate.lua [hunk 1]
```diff
@@ -156,13 +144,15 @@ local options, args = {
...
     include_friendly = false,
+    maximum = -1,
 }, {...}
 
 local positionals = argparse.processArgsGetopt(args, {
     {'h', 'help', handler = function() options.help = true end},
     {'m', 'method', handler = function(arg) options.method = killMethod[arg:upper()] end, hasArg = true},
     {'o', 'only-visible', handler = function() options.only_visible = true end},
     {'f', 'include-friendly', handler = function() options.include_friendly = true end},
+    {'x', 'maximum', handler = function(arg) options.maximum = tonumber(arg) end, hasArg = true},
```
Comment (myk002) [inline exterminate.lua:155]: use `argparse.positiveInt` to sanity check format (or `argparse.nonnegativeInt` if you want to allow `0`). if the player passes `-x many`, then `options.maximum` will be set to `nil` and we'll get an exception in the `options.maximum > 0` check below.

**Review by myk002 — COMMENTED** [review]
please add documentation to `docs/exterminate.rst` and `changelog.txt`

File: exterminate.lua [hunk 2]
```diff
@@ -109,6 +95,8 @@ function killUnit(unit, method)
     elseif method == killMethod.VAPORIZE then
         vaporizeUnit(unit)
     elseif method == killMethod.DISINTEGRATE then
+        -- Teleport them to world height. Out of sight, out of mind.
+        dfhack.units.teleport(unit, xyz2pos(unit.pos.x, unit.pos.y, df.global.world.map.z_count-1))
```
Comment (myk002) [inline exterminate.lua:99]: does it work if we instead mark them as hidden (in ambush)? it seems like teleporting has more things that can go wrong (e.g. what if your castle in the sky happens to have a kinder care center at that spot?)

Comment (Crystalwarrior) [reply exterminate.lua:99]: The idea is to prevent smoke clouds as well, maybe teleport them to Hell?

File: changelog.txt [hunk 3]
```diff
@@ -73,6 +73,9 @@ Template for new versions:
 - `gui/launcher`: refresh default tag filter when mortal mode is toggled in `gui/control-panel` so changes to which tools autocomplete take effect immediately
 - `gui/civ-alert`: you can now register multiple burrows as civilian alert safe spaces
 - `exterminate`: add ``all`` target for convenient scorched earth tactics
+- `exterminate`: add ``--max`` option to limit number of exterminated creatures
+- `exterminate`: add ``--knockout`` method for a non-lethal "extermination"
+- `exterminate`: add ``all`` target for convenient scorched earth tactics
```
Comment (myk002) [inline changelog.txt:78]:
```suggestion
```

File: changelog.txt [hunk 4]
```diff
@@ -73,6 +73,9 @@ Template for new versions:
 - `gui/launcher`: refresh default tag filter when mortal mode is toggled in `gui/control-panel` so changes to which tools autocomplete take effect immediately
 - `gui/civ-alert`: you can now register multiple burrows as civilian alert safe spaces
 - `exterminate`: add ``all`` target for convenient scorched earth tactics
+- `exterminate`: add ``--max`` option to limit number of exterminated creatures
```
Comment (myk002) [inline changelog.txt:76]:
```suggestion
- `exterminate`: add ``--limit`` option to limit number of exterminated creatures
```

File: docs/exterminate.rst [hunk 5]
```diff
@@ -48,6 +48,8 @@ Options
     on the map.
 ``-f``, ``--include-friendly``
     Specifies the tool should also kill units friendly to the player.
+``-x``, ``--max <num>``
```
Comment (myk002) [inline docs/exterminate.rst:51]:
```suggestion
``-l``, ``--limit <num>``
```

File: docs/exterminate.rst [hunk 6]
```diff
@@ -64,6 +66,7 @@ Methods
 :magma: Boil the unit in magma (not recommended for magma-safe creatures).
 :butcher: Will mark the units for butchering instead of killing them. This is
     more useful for pets than armed enemies.
+:knockout: Will put units into an unconscious state for 30k ticks.
```
Comment (myk002) [inline docs/exterminate.rst:69]:
```suggestion
:knockout: Will put units into an unconscious state for 30k ticks (about a month).
```

File: exterminate.lua [hunk 7]
```diff
@@ -25,11 +25,21 @@ local function isUnitFriendly(unit)
     if dfhack.units.isDanger(unit) then
         return false
     end
+    local isInAdvGroup = false
+    local isAdvPet = false
+    local adv = dfhack.world.getAdventurer()
+    if adv then
+        isInAdvGroup = unit.relationship_ids.GroupLeader == dfhack.world.getAdventurer().id
+        isAdvPet = unit.relationship_ids.PetOwner == dfhack.world.getAdventurer().id
+    end
```
Comment (myk002) [inline exterminate.lua:34]:
```suggestion
    local adv = dfhack.world.getAdventurer()
    if adv then
        if adv == unit or
            unit.relationship_ids.GroupLeader == adv.id or
            unit.relationship_ids.PetOwner == adv.id
        then
            return true
        end
    end
```

File: exterminate.lua [hunk 8]
```diff
@@ -25,11 +25,21 @@ local function isUnitFriendly(unit)
...
     return dfhack.units.isOwnCiv(unit) or
         dfhack.units.isOwnGroup(unit) or
         dfhack.units.isVisiting(unit) or
         dfhack.units.isTame(unit) or
-        dfhack.units.isDomesticated(unit)
+        dfhack.units.isDomesticated(unit) or
+        unit == adv or
+        isInAdvGroup or
+        isAdvPet
```
Comment (myk002) [inline exterminate.lua:42]:
```suggestion
        dfhack.units.isDomesticated(unit)
```

File: exterminate.lua [hunk 9]
```diff
@@ -25,11 +25,21 @@ local function isUnitFriendly(unit)
...
     end
+    local isInAdvGroup = false
+    local isAdvPet = false
+    local adv = dfhack.world.getAdventurer()
+    if adv then
+        isInAdvGroup = unit.relationship_ids.GroupLeader == dfhack.world.getAdventurer().id
+        isAdvPet = unit.relationship_ids.PetOwner == dfhack.world.getAdventurer().id
+    end
     return dfhack.units.isOwnCiv(unit) or
         dfhack.units.isOwnGroup(unit) or
```
Comment (myk002) [inline exterminate.lua:36]: how does isOwnGroup behave in adventure mode?

File: exterminate.lua [hunk 10]
```diff
@@ -156,13 +161,15 @@ local options, args = {
     method = killMethod.INSTANT,
     only_visible = false,
     include_friendly = false,
+    maximum = -1,
```
Comment (myk002) [inline exterminate.lua:164]:
```suggestion
    limit = -1,
```

File: exterminate.lua [hunk 11]
```diff
@@ -156,13 +161,15 @@ local options, args = {
...
 }, {...}
 
 local positionals = argparse.processArgsGetopt(args, {
     {'h', 'help', handler = function() options.help = true end},
     {'m', 'method', handler = function(arg) options.method = killMethod[arg:upper()] end, hasArg = true},
     {'o', 'only-visible', handler = function() options.only_visible = true end},
     {'f', 'include-friendly', handler = function() options.include_friendly = true end},
+    {'x', 'maximum', handler = function(arg) options.maximum = tonumber(arg) end, hasArg = true},
```
Comment (myk002) [inline exterminate.lua:172]:
```suggestion
    {'l', 'limit', handler = function(arg) options.limit = argparse.positiveInt(arg, 'limit') end, hasArg = true},
```

File: exterminate.lua [hunk 12]
```diff
@@ -217,7 +224,9 @@ elseif positionals[1]:split(':')[1] == "all" then
     local selected_caste = positionals[1]:split(':')[2]
 
     for _, unit in ipairs(df.global.world.units.active) do
-
+        if options.maximum > 0 and count >= options.maximum then
```
Comment (myk002) [inline exterminate.lua:227]:
```suggestion
        if options.limit > 0 and count >= options.limit then
```

File: exterminate.lua [hunk 13]
```diff
@@ -269,6 +278,9 @@ else
     target = selected_race
 
     for _, unit in pairs(df.global.world.units.active) do
+        if options.maximum > 0 and count >= options.maximum then
```
Comment (myk002) [inline exterminate.lua:281]:
```suggestion
        if options.limit > 0 and count >= options.limit then
```

File: exterminate.lua (hunk 9 above)
Comment (Crystalwarrior) [reply exterminate.lua:36]: it doesn't. isOwnGroup only checks plotinfo which is irrelevant to adventure mode

File: docs/exterminate.rst (hunk 6 above)
Comment (Crystalwarrior) [reply docs/exterminate.rst:69]: more testing needed for this one, I dunno if adventure mode ticks align with fortress mode ticks

File: exterminate.lua (hunk 9 above)
Comment (myk002) [reply exterminate.lua:39]: does that mean it always returns false, or that it returns garbage data?

File: docs/exterminate.rst (hunk 6 above)
Comment (myk002) [reply docs/exterminate.rst]: they don't. the month estimate is for fort mode. you could define either define a time span that `knockout` will last and adjust different ticks based on the mode, or else document the different time spans that `knockout` will last in fort mode vs. adventure mode.

**Comment by myk002** [comment]
note that you can add the code change suggestions to a batch and commit them directly instead of reimplementing the suggestions yourself. That also has the benefit of resolving the suggestion.

File: changelog.txt [hunk 14]
```diff
@@ -73,6 +73,8 @@ Template for new versions:
 - `gui/launcher`: refresh default tag filter when mortal mode is toggled in `gui/control-panel` so changes to which tools autocomplete take effect immediately
 - `gui/civ-alert`: you can now register multiple burrows as civilian alert safe spaces
 - `exterminate`: add ``all`` target for convenient scorched earth tactics
+- `exterminate`: add ``--limit`` option to limit number of exterminated creatures
+- `exterminate`: add ``--knockout`` method for a non-lethal "extermination"
```
Comment (myk002) [inline changelog.txt:77]:
```suggestion
- `exterminate`: add ``knockout`` and ``traumatize`` method for a non-lethal incapacitation
```

**Review by myk002 — APPROVED** [review]

**Comment by myk002** [comment]
I determined that `dfhack.units.isOwnGroup` will return garbage data in adventure mode, but it's due to a DF bug. I'll take that up with Putnam.

**Pull request merged**